### PR TITLE
Throw SecurityTokenSignatureValidationFailedException when none of the provided keys verify the signature

### LINKED ISSUE #1268: Throw SecurityTokenSignatureKeyNotFoundException when none of the supplied keys verify the signature.
If there are keys made available to the runtime and none of them result in a verified signature, then throw SecurityTokenSignatureKeyNotFoundException. This will allow asp.net to trigger a refresh of metadata. 

Currently we do not throw if a kid is not found in the inbound jwt.

see: https://github.com/aspnet/AspNetCore/issues/12310

### PR DESCRIPTION
**Pull request by RojaEnnam**
Fixes  #1268

- Throw `SecurityTokenSignatureKeyNotFoundException` when a security token does not contain a key identifier and the signature validation is not successful with keys found by the runtime.
- Modify current logic to also resolve encryption keys when security token does not contain key identifier.

### CODE REVIEW

**Comment by RojaEnnam** [comment]
The duplicated code is not all from this PR. I looked into reformatting ValidateSignature method in Saml/Saml2SecurityTokenHandlers. One thing I found was Saml/Saml2Assertion does not have a base class.

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs [hunk 1]
```diff
@@ -96,7 +96,7 @@ internal static class LogMessages
         // 10500 - SignatureValidation
         public const string IDX10500 = "IDX10500: Signature validation failed. No security keys were provided to validate the signature.";
         public const string IDX10501 = "IDX10501: Signature validation failed. Unable to match key: \nkid: '{0}'.\nExceptions caught:\n '{1}'. \ntoken: '{2}'.";
-        public const string IDX10503 = "IDX10503: Signature validation failed. Keys tried: '{0}'.\nExceptions caught:\n '{1}'.\ntoken: '{2}'.";
+        public const string IDX10503 = "IDX10503: Signature validation failed. Token does not have a kid. Keys tried: '{0}'.\nExceptions caught:\n '{1}'.\ntoken: '{2}'.";
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/LogMessages.cs:99]: This doesn't work for SAML/SAML2

Comment (RojaEnnam) [reply src/Microsoft.IdentityModel.Tokens/LogMessages.cs:99]: You're right. I missed that. I made the change.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs [hunk 2]
```diff
@@ -1104,7 +1104,7 @@ private SamlSecurityToken ValidateSignature(SamlSecurityToken samlToken, string
                 throw LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(FormatInvariant(TokenLogMessages.IDX10501, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs]: We could change message IDX10501 to be specific for SAML and use 'KeyInfo' as you did with IDX10512

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs [hunk 3]
```diff
@@ -105,6 +105,7 @@ internal static class LogMessages
         public const string IDX10509 = "IDX10509: Signature validation failed. The user defined 'Delegate' specified in TokenValidationParameters did not return a '{0}', but returned a '{1}' when reading token: '{2}'.";
         public const string IDX10510 = "IDX10510: Signature validation failed. The user defined 'Delegate' specified in TokenValidationParameters returned null when reading token: '{0}'.";
         public const string IDX10511 = "IDX10511: Signature validation failed. Keys tried: '{0}'. \nkid: '{1}'. \nExceptions caught:\n '{2}'.\ntoken: '{3}'.";
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/LogMessages.cs:107]: Change 'kid' to 'KeyInfo'

**Comment by brentschmaltz** [comment]
@RojaEnnam looks good, only change i have to suggest is to use 'KeyInfo' instead of 'kid' for saml tokens.
The reason that there is no base class for Saml1 and Saml2 is that they are very different objects.

We could have ValidateSignature across all tokens perhaps in the future.

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs (hunk 3 above)
Comment (RojaEnnam) [reply src/Microsoft.IdentityModel.Tokens/LogMessages.cs:107]: IDX10511 is not used for saml. For IDX10501 I added IDX10513 to be saml specific.

**Review by brentschmaltz — APPROVED** [review]

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs (hunk 3 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Tokens/LogMessages.cs:107]: Looks like 10503 should have kid -> KeyInfo then?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs [hunk 4]
```diff
@@ -444,10 +444,10 @@ private Saml2SecurityToken ValidateSignature(Saml2SecurityToken samlToken, strin
 
             //if there was a key match with what was found in tokenValidationParameters most likely metadata is stale. throw SecurityTokenSignatureKeyNotFoundException
             if (!keyMatched && canMatchKey && keysAttempted.Length > 0)
-                throw LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(FormatInvariant(TokenLogMessages.IDX10501, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
+                throw LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(FormatInvariant(TokenLogMessages.IDX10513, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
 
             if (keysAttempted.Length > 0)
-                throw LogExceptionMessage(new SecurityTokenInvalidSignatureException(FormatInvariant(TokenLogMessages.IDX10503, keysAttempted, exceptionStrings, samlToken)));
+                throw LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(FormatInvariant(TokenLogMessages.IDX10512, keysAttempted, exceptionStrings, samlToken)));
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs:454]: Is this logically correct still?

If we look at the test cases, we are raising KeyNotFound for cases where attribute or signatures have been tampered with

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs (hunk 3 above)
Comment (RojaEnnam) [reply src/Microsoft.IdentityModel.Tokens/LogMessages.cs:107]: 10503  is used in M.IM.JsonWebTokens and S.IM.T.jwt, saml specific for 10503 is 10512.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs (hunk 4 above)
Comment (RojaEnnam) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs:454]: I was wondering about that myself. Upon more testing I think its best to match saml logic to jwt. 
I went ahead made the changes in logic and tests.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs [hunk 5]
```diff
@@ -1099,14 +1099,20 @@ private SamlSecurityToken ValidateSignature(SamlSecurityToken samlToken, string
...
 
-            // if there was a keymatch with what was found in tokenValidationParameters most likely metadata is stale. throw SecurityTokenSignatureKeyNotFoundException
-            if (!keyMatched && canMatchKey && keysAttempted.Length > 0)
-                throw LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(FormatInvariant(TokenLogMessages.IDX10501, samlToken.Assertion.Signature.KeyInfo, exceptionStrings, samlToken)));
-
-            if (keysAttempted.Length > 0)
-                throw LogExceptionMessage(new SecurityTokenInvalidSignatureException(FormatInvariant(TokenLogMessages.IDX10503, keysAttempted, exceptionStrings, samlToken)));
-
-            throw LogExceptionMessage(new SecurityTokenInvalidSignatureException(TokenLogMessages.IDX10500));
+            if (canMatchKey)
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.cs:1102]: This could be simplified (a lot of the `else`s are redundant)

**Review by keegan-caruso — APPROVED** [review]

**Pull request merged**